### PR TITLE
Additional checks on Marlin/Darlin proof and vk + utility function to compute proof size given free parameters

### PR DESCRIPTION
**Pull request by 95DDB**

### CODE REVIEW

**Comment by 95DDB** [comment]
Closed and moved to https://github.com/HorizenOfficial/zendoo-cctp-lib/pull/19

File: proof-systems/src/darlin/pcd/simple_marlin.rs [hunk 1]
```diff
@@ -23,6 +23,143 @@ use std::marker::PhantomData;
...
+pub fn is_vk_consistent_with<G: AffineCurve, D: Digest>(
+    vk: &MarlinVerifierKey<G::ScalarField, InnerProductArgPC<G, D>>,
+    segment_size: usize,
+    max_domain_h_size: usize,
+    max_domain_k_size: usize,
+) -> bool
+{
+    let num_constraints = vk.index_info.num_constraints.next_power_of_two();
+    let num_variables = vk.index_info.num_variables.next_power_of_two();
+    let domain_h_size = std::cmp::max(num_constraints, num_variables);
```
Comment (UlrichHaboeck75) [inline proof-systems/src/darlin/pcd/simple_marlin.rs:37]: Even if it doesn't change the logic, I'd use the next_power_of_two() to define domain_h_size. 
(For now, a domain size is always two-adic).

File: proof-systems/src/darlin/pcd/simple_marlin.rs [hunk 2]
```diff
@@ -23,6 +23,143 @@ use std::marker::PhantomData;
...
+) -> bool
+{
+    let segment_size: usize = 1 << proof.pc_proof.proof.l_vec.len();
+    let zk_bound = if proof.pc_proof.proof.hiding_comm.is_some() { 1 } else { 0 };
+    let w_z_a_b_segs = ((max_domain_h_size + 2 * zk_bound) as f64/segment_size as f64).ceil() as usize;
+    let t_segs = ((max_domain_h_size as f64/segment_size as f64)).ceil() as usize;
+    let z_1_segs = ((max_domain_h_size + 3 * zk_bound) as f64/segment_size as f64).ceil() as usize;
+    let h_1_segs = ((2 * max_domain_h_size + 4 * zk_bound - 2) as f64/segment_size as f64).ceil() as usize;
+    let z_2_segs = (max_domain_k_size as f64/segment_size as f64).ceil() as usize;
+    let h_2_segs =  ((3 * max_domain_k_size - 3) as f64/segment_size as f64).ceil() as usize;
```
Comment (UlrichHaboeck75) [inline proof-systems/src/darlin/pcd/simple_marlin.rs:64]: Why not choose the claimed/derived domain_h_size from the vk?

File: proof-systems/src/darlin/pcd/simple_marlin.rs [hunk 3]
```diff
@@ -23,6 +23,143 @@ use std::marker::PhantomData;
...
+
+/// Given segment_size, density, zk, proof_type, return the maximum value of |H| s.t. proof size will be <= max_proof_size
+/// and the corresponding proof size and vk size
+pub fn compute_max_domain_h_size(
+    segment_size: usize,
+    density: usize,
+    zk: bool,
+    max_proof_size: usize,
+    proof_type: &str
+) -> (usize, usize, usize)
```
Comment (UlrichHaboeck75) [inline proof-systems/src/darlin/pcd/simple_marlin.rs:87]: I'd still opt for a ``compute_max_constraints_and_variables()`` instead, as it is might be more informative in practice:
Given the density, a circuit might have more constraints than the one returned by the function. The reason for this is that the function applies the density to a power of two, which is only realistic if the circuit does in fact use all "space" given by a two-adic domain. In practice, the constraint system is padded, but this padding does not affect the non-zero entries hence the domain K needed.

File: proof-systems/src/darlin/pcd/simple_marlin.rs [hunk 4]
```diff
@@ -23,6 +23,143 @@ use std::marker::PhantomData;
...
+        let t_segs = ((h as f64/segment_size as f64)).ceil() as usize;
+        let z_1_segs = ((h + 3 * zk_bound) as f64/segment_size as f64).ceil() as usize;
+        let h_1_segs = ((2 * h + 4 * zk_bound - 2) as f64/segment_size as f64).ceil() as usize;
+        let z_2_segs = (k as f64/segment_size as f64).ceil() as usize;
+        let h_2_segs =  ((3 * k - 3) as f64/segment_size as f64).ceil() as usize;
+
+        if w_z_a_b_segs > 255 || t_segs > 255 || z_1_segs > 255 || h_1_segs > 255 || z_2_segs > 255 || h_2_segs > 255 {
+            max_supported_h_size += 1;
+            continue;
+        }
```
Comment (UlrichHaboeck75) [inline proof-systems/src/darlin/pcd/simple_marlin.rs:108]: Given a fixed segment size, why do we INCREASE ``max_supported_h_size`` if one of the polynomials already exceeds the sanity bound of 255 segments? For higher domain sizes this clause will be triggered, too.

File: proof-systems/src/darlin/tests/final_darlin.rs [hunk 5]
```diff
@@ -369,6 +370,15 @@ pub fn generate_test_data<'a, G1: AffineCurve, G2: AffineCurve, D: Digest + 'a,
         info.clone()
     ).unwrap();
 
+    let domain_h_size = std::cmp::max(index_vk.index_info.num_constraints, index_vk.index_info.num_variables).next_power_of_two();
+    let domain_k_size = index_vk.index_info.num_non_zero.next_power_of_two();
+    let segment_size = (segment_size - 1).next_power_of_two();
```
Comment (UlrichHaboeck75) [inline proof-systems/src/darlin/tests/final_darlin.rs:375]: Why segment_size-1? I'd expect the segment_size to be the size, and not the supported degree.

File: proof-systems/src/darlin/tests/final_darlin.rs [hunk 6]
```diff
@@ -369,6 +370,15 @@ pub fn generate_test_data<'a, G1: AffineCurve, G2: AffineCurve, D: Digest + 'a,
...
+    assert!(index_vk.is_valid());
+    assert!(is_vk_consistent_with(&index_vk, segment_size, domain_h_size, domain_k_size));
+    assert!(!is_vk_consistent_with(&index_vk, segment_size * 2, domain_h_size, domain_k_size));
+    assert!(!is_vk_consistent_with(&index_vk, segment_size, domain_h_size/2, domain_k_size));
+    assert!(!is_vk_consistent_with(&index_vk, segment_size, domain_h_size, domain_k_size/2));
```
Comment (UlrichHaboeck75) [inline proof-systems/src/darlin/tests/final_darlin.rs:380]: why checking a self generated vk?

File: proof-systems/src/darlin/tests/simple_marlin.rs [hunk 7]
```diff
@@ -148,6 +148,15 @@ pub fn generate_test_data<'a, G: AffineCurve, D: Digest + 'a, R: RngCore>(
         &committer_key, circ.clone()
     ).unwrap();
 
+    let domain_h_size = std::cmp::max(index_vk.index_info.num_constraints, index_vk.index_info.num_variables).next_power_of_two();
+    let domain_k_size = index_vk.index_info.num_non_zero.next_power_of_two();
+    let segment_size = (segment_size - 1).next_power_of_two();
```
Comment (UlrichHaboeck75) [inline proof-systems/src/darlin/tests/simple_marlin.rs:153]: same here: why minus 1?

**Review by UlrichHaboeck75 — COMMENTED** [review]
I'd recommend to do a ``get_max_constraints_and_variables()`` instead of the present ``get_max_domain_h_size()``, see my comment for details.

**Pull request closed without merging**